File: 1207-delete-nodes-and-return-forest/1207-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<TreeNode> ans = new ArrayList<>();
    private boolean del[] = new boolean[1001];
    private TreeNode dfs(TreeNode root) {
        if (root == null) return root;

        root.left = dfs(root.left);
        root.right = dfs(root.right);

        if (del[root.val]) {
            if (root.left != null && !del[root.left.val]) ans.add(root.left);
            if (root.right != null && !del[root.right.val]) ans.add(root.right);
            return null;
        }

        return root;
    }
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        if (root == null) return ans;
        for (int d: to_delete) del[d] = true;
        if (!del[root.val]) ans.add(root);
        dfs(root);
        return ans;
    }
}

File: 1207-delete-nodes-and-return-forest/1207-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<TreeNode*> ans;
    bool del[1001] = { 0 };
    TreeNode* dfs(TreeNode* root, TreeNode* prev) {
        if (!root) return nullptr;

        root->left = dfs(root->left, root);
        root->right = dfs(root->right, root);

        if (del[root->val]) {
            if (root->left && !del[root->left->val]) ans.push_back(root->left);
            if (root->right && !del[root->right->val]) ans.push_back(root->right);
            return nullptr;
        }

        return root;
    }
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        if (!root) return {};
        for (int d: to_delete) del[d] = true;
        if (!del[root->val]) ans.push_back(root);
        dfs(root, nullptr);
        return ans;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();